src/main/java/com/celements/navigation/cmd/GetNotMappedMenuItemsForParentCommand.java

```diff
@@ -35,31 +37,64 @@
 import com.xpn.xwiki.XWikiException;
 import com.xpn.xwiki.doc.XWikiDocument;
 import com.xpn.xwiki.objects.BaseObject;
+import com.xpn.xwiki.web.Utils;
 
 public class GetNotMappedMenuItemsForParentCommand {
 
-  private static Log LOGGER = LogFactory.getFactory().getInstance(
+  private static Logger LOGGER = LoggerFactory.getLogger(
       GetNotMappedMenuItemsForParentCommand.class);
 
-  private HashMap<String, Map<String, List<TreeNode>>> menuItems;
+  private final Map<String, Map<String, List<TreeNode>>> menuItems;
   
   private int queryCount = 0;
 
+  private XWikiContext getContext() {
+    return (XWikiContext) Utils.getComponent(Execution.class).getContext(
+        ).getProperty("xwikicontext");
+  }
+
   public GetNotMappedMenuItemsForParentCommand() {
-    menuItems = new HashMap<String, Map<String, List<TreeNode>>>();
+    /* ConcurrentHashMap is needed to prevent:
+     * 1) ConcurrentModificationException if Thread A iterates over menuItems and
+     *    Thread B is changing it.
+     * 2) Thread B changes the HashMap and Thread B does not synchronize on read
+     *    which may lead that it reads an old version from its memory cache.
+     *    Details see JSR133
+     */
```

Comment:
1) menuItems is never being iterated, however accessing menuItems is sufficient as a requirement for using ConcurrentHashMap
2) The second thread should be called Thread A (2x B in text)